go-controller/pkg/ovn/service.go

```diff
@@ -228,6 +236,10 @@ func (ovn *Controller) createService(service *kapi.Service) error {
 					// Skip creating LB if endpoints watcher already did it
 					if _, hasEps := ovn.getServiceLBInfo(loadBalancer, vip); hasEps {
 						klog.V(5).Infof("Load Balancer already configured for %s, %s", loadBalancer, vip)
+					} else if ep != nil {
+						if err := ovn.AddEndpoints(ep); err != nil {
```

Comment:
we shouldn't be adding endpoints here. Endpoints will be handled by the endpoints watcher. It just so happens that the function in add endpoints will also create the service (due to the relationship of service and endpoints). line 236-237 already attempts to skip LB creation if it sees that endpoints has already done the job. Is this commit saying that method is not working?

Reply:
Actually I left a comment regarding the reason for this (https://github.com/ovn-org/ovn-kubernetes/pull/1306#issuecomment-622369501). We can get a endpoint creation before the service creation, in which case we end up with a "ACL reject", whereas we should have a normal functioning service.

Reply:
But in that case we don't solve this problem? In case the endpoint creation reaches us before the service one, we don't do anything (because we find no service). And then later once the service creation comes: even if we see that there is an endpoint, we still don't treat it. Meaning the ACL reject is still there.

Reply:
might be easier to just sync up tmrw on this, but the endpoint creation doesn't check for the service map, it actually goes and gets the service from k8s. This is how it was working before I added the service stuff. Endpoints is actually responsible for configuring the load balancer. The service watcher is actually only needed for services without endpoints. Let's talk tmrw and I can take a closer look then too.

Reply:
> it actually goes and gets the service from k8s.

I think this is where the problem lies: `ovn.watchFactory.GetService(ep.Namespace, ep.Name)` it gets it from the api-server's cache store (not etcd). In case that the api-server's cache has not been populate with the service once we run `AddEndpoint` that logic falls apart. But yes, let's sync tomorrow on this.

Reply:
Alex and I spoke offline about this. There is a chance with today's informer implementation that the cache is out of sync and this race can happen. We also looked at the potential path of both watchers simultaneously calling AddEndpoints, but it should be OK because we have proper locking and config setting, it should just result in one extra ovn-nbctl config.